llms/hf_llm/utils.py

```diff
@@ -0,0 +1,146 @@
+import glob
+import json
+import logging
+from pathlib import Path
+from typing import Generator, Tuple
+
+import mlx.core as mx
+import mlx.nn as nn
+import models.llama as llama
+import models.phi2 as phi2
+from huggingface_hub import snapshot_download
+
+# Local imports
+from models.base import BaseModelArgs
+from transformers import AutoTokenizer, PreTrainedTokenizer
+
+# Constants
+MODEL_MAPPING = {
+    "LlamaForCausalLM": llama,
+    "MistralForCausalLM": llama,  # mistral is compatible with llama
+    "PhiForCausalLM": phi2,
+}
+
+
+def _get_classes(config: BaseModelArgs):
+    """
+    Retrieve the model and model args classes based on the configuration.
+
+    Args:
+        config (ModelArgs): The model configuration.
+
+    Returns:
+        A tuple containing the Model class and the ModelArgs class, or None if not found.
+    """
+    architectures = config.get("architectures", [])
+
+    for arch in architectures:
+        if arch in MODEL_MAPPING:
+            return MODEL_MAPPING[arch].Model, MODEL_MAPPING[arch].ModelArgs
+    return None
+
+
+def get_model_path(path_or_hf_repo: str) -> Path:
+    """
+    Ensures the model is available locally. If the path does not exist locally,
+    it is downloaded from the Hugging Face Hub.
+
+    Args:
+        path_or_hf_repo (str): The local path or Hugging Face repository ID of the model.
+
+    Returns:
+        Path: The path to the model.
+    """
+    model_path = Path(path_or_hf_repo)
+    if not model_path.exists():
+        model_path = Path(
+            snapshot_download(
+                repo_id=path_or_hf_repo,
+                allow_patterns=["*.json", "*.safetensors", "*.py", "tokenizer.model"],
```

Comment:
Include the `*.py` file to download patterns. In case a model has a custom config mapping, it will throw an error at `autoConfig.from_pretrained(model_path)`. user will still be prompted whether they want to trust remote code when loading the model/config from_pretrained.